src/core/namespace.c

```diff
@@ -1576,13 +1588,23 @@ int setup_namespace(
                         n_mount_images,
                         tmp_dir, var_tmp_dir,
                         creds_path,
-                        log_namespace);
+                        log_namespace,
+                        setup_propagate,
+                        root_is_read_only);
 
         if (n_mounts > 0) {
                 m = mounts = new0(MountEntry, n_mounts);
                 if (!mounts)
                         return -ENOMEM;
 
+                /* We need at least /run to be writable, so add a tmpfs in case the user doesn't */
+                if (setup_propagate && root_is_read_only) {
+                        *(m++) = (MountEntry) {
+                                .path_const = "/run",
+                                .mode = TMPFS,
+                        };
+                }
```

Comment:
hmm, i am not sure how i feel about this. I have the suspicion we should solve this differently.

I.e. i think this should be bound to MountAPIVFS= in some way, and because this is generally useful. In fact MountAPIVFS=yes should probably have done this by default anyway since forever.

So maybe we should change this now, even if that's an api break of some kind. i.e. MountAPIVFS=yes should from now on mount a fresh /run too, and we should introduce MountAPIVFS=restricted or so, which is a special option to get the status quo ante back, i.e. where it only mounts /proc + /dev + /sys. Then announce that in NEWS as an incompatibility.

In fact, we probably should go further and declare that the setup logic of the propagation/incoming dirs is also part of MountAPIVFS=yes behaviour?

Reply:
mmmh we can't change MountAPIVFS's to add a new value, as the DBUS interface is a boolean :-/

IMHO we can just enable /run as a tmpfs unconditionally. I mean, loads of stuff these days would be completely borken with a read-only /run. wdyt?

Reply:
also a tmpfs has low prio, so if a user passes something through, it wins.

Reply:
@keszybz suggested adding a MountAPIVFS2 DBUS interface. The config stuff could switch to that, and we keep the old interface for backward compat. This is certainly an option - however, I'm a bit wary of adding new interfaces if there's not a real need for it, given we have to commit (and never change them). I pushed a commit to do the change unconditionally and document it - given we are at the start of the cycle, we could add MountAPIVFS2 only later if we figure out it was actually needed after all. Does this sound reasonable?

Reply:
I think I was a bit optimistic. In normal cases (no RootDirectory, no RootImage) you get /run from the host. Always adding a tmpfs means a lot of API entry points, like sockets, are not available and have to be manually specified.
I've changed the diff to automatically mount the dbus, notify and journal sockets as that's the bare minimum - but maybe it's worth reconsidering and going back to the original proposal of adding a tmpfs only if RootDirectory/RootImage are reado-only?

Reply:
So as mentioned privately I misunderstood the suggestion and I've now reworked it: a tmpfs is added on /run unconditionally with respect to rootimage/rootdir/readwrite, but only if /run is not already a mount point. Things seem to be working a bit better :-)

I agree on making the propagation/incoming bits part of the APIVFS behaviour/spec, and I'll do that next.